pydantic/types.py

```diff
@@ -442,77 +443,20 @@ def __eq__(self, other: Any) -> bool:
 SecretType = TypeVar('SecretType', str, bytes)
 
 
-class SecretField(abc.ABC, Generic[SecretType]):
-    _error_kind: str
-
+class SecretField(Generic[SecretType]):
     def __init__(self, secret_value: SecretType) -> None:
         self._secret_value: SecretType = secret_value
 
     def get_secret_value(self) -> SecretType:
         return self._secret_value
 
     @classmethod
-    def __get_pydantic_core_schema__(cls, source: type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
-        validator = SecretFieldValidator(cls)
-        js_modify_functions: list[Any] = []
-        if issubclass(cls, SecretStr):
-            # Use a lambda here so that `apply_metadata` can be called on the validator before the override is generated
-            def js_cs_override1(_c: Any, h: Any) -> Any:
-                return h(core_schema.str_schema(min_length=validator.min_length, max_length=validator.max_length))
-
-            js_modify_functions.append(js_cs_override1)
-
-        elif issubclass(cls, SecretBytes):
-
-            def js_cs_override2(_c: Any, h: Any) -> Any:
-                return h(core_schema.bytes_schema(min_length=validator.min_length, max_length=validator.max_length))
-
-            js_modify_functions.append(js_cs_override2)
-
-        metadata = build_metadata_dict(
-            cs_update_function=validator.__pydantic_update_schema__, js_functions=js_modify_functions
-        )
-        return core_schema.general_after_validator_function(
-            validator,
-            core_schema.union_schema(
-                [core_schema.is_instance_schema(cls), cls._pre_core_schema()],
-                strict=True,
-                custom_error_type=cls._error_kind,
-            ),
-            metadata=metadata,
-            serialization=core_schema.plain_serializer_function_ser_schema(
-                cls._serialize, info_arg=True, json_return_type='str'
-            ),
-        )
-
-    @classmethod
-    def _serialize(
-        cls, value: SecretField[SecretType], info: core_schema.SerializationInfo
-    ) -> str | SecretField[SecretType]:
-        if info.mode == 'json':
-            # we want the output to always be string without the `b'` prefix for byties,
-            # hence we just use `secret_display`
-            return secret_display(value)
-        else:
-            return value
-
-    @classmethod
-    @abc.abstractmethod
-    def _pre_core_schema(cls) -> core_schema.CoreSchema:
-        ...
-
-    @classmethod
-    def __get_pydantic_json_schema__(
-        cls, core_schema: core_schema.CoreSchema, handler: GetJsonSchemaHandler
-    ) -> JsonSchemaValue:
-        field_schema = handler(core_schema)
-        update_not_none(
-            field_schema,
-            type='string',
-            writeOnly=True,
-            format='password',
-        )
-        return field_schema
+    def __prepare_pydantic_annotations__(cls, source: type[Any], annotations: Iterator[Any]) -> Iterator[Any]:
```

Comment:
```suggestion
    def __prepare_pydantic_annotations__(cls, source: type[Any], annotations: Iterable[Any]) -> Iterable[Any]:
```
I saw `res = list(prepare(source_type, tuple(annotations)))` elsewhere so I don't think the thing getting passed into  this function is an Iterator in general

Reply:
Yup thanks.

I think ideally we want to encourage the signature you are suggesting for users.
As part of writing the docs I plan on making some protocol classes that describe these APIs.
They will be helpful in terms of discoverability, reducing boilerplate because you can add it as a temporary base class and get autocomplete but also maybe leave it there permanently.